### PR TITLE
[Stats] Fix harvestor threads + Fix flaky stats shutdown.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

1. stats shutdown has a deadlock condition. It is fixed by this PR. (We hold a lock when we join threads, which also need to be held inside threads).
2. Cleanup harvestor threads.

There's one ugly part here. We don't hold locks when we access threads variables (e.g., `t_` or `harvestor_thread_`) because `ray::stats::shutdown` guarantees that these methods are called only by a single thread. When I added locks around, there was always deadlock (idk the exact cause). 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Review by edoakes — APPROVED** [review]

File: thirdparty/patches/opencensus-cpp-shutdown-api.patch [hunk 1]
```diff
@@ -24,49 +100,43 @@ diff --git opencensus/stats/internal/stats_exporter.cc b/opencensus/stats/intern
...
+     // SleepFor() returns immediately when given a negative duration.
+     absl::SleepFor(next_export_time - absl::Now());
+     // In case the last export took longer than the export interval, we
+@@ -114,6 +134,12 @@ void StatsExporterImpl::RunWorkerLoop() {
  // StatsExporter
  // -------------
  
 +void StatsExporter::Shutdown() {
 +  StatsExporterImpl::Get()->Shutdown();
++  StatsExporter::ExportForTesting();
```
Comment (ashione) [inline thirdparty/patches/opencensus-cpp-shutdown-api.patch:119]: Why we need do exporting in shutdown? we can't guarantee all metric data will be flushed.

File: src/ray/stats/stats.h [hunk 2]
```diff
@@ -109,6 +109,7 @@ static inline void Shutdown() {
   }
   metrics_io_service_pool->Stop();
   opencensus::stats::StatsExporter::Shutdown();
+  opencensus::stats::DeltaProducer::Get()->Shutdown();
```
Comment (ashione) [inline src/ray/stats/stats.h:112]: Should `DeltaProducer` shutdown before StatsExporter::Shutdown?

File: thirdparty/patches/opencensus-cpp-shutdown-api.patch (hunk 1 above)
Comment (rkooo567) [reply thirdparty/patches/opencensus-cpp-shutdown-api.patch:119]: Ah, good point. Maybe I should add flush & export for testing in ray shutdown to make sure all remaining metrics are exported?

File: src/ray/stats/stats.h (hunk 2 above)
Comment (rkooo567) [reply src/ray/stats/stats.h:112]: so that we can flush and export right? I think that's a good idea.

**Comment by rkooo567** [comment]
@ashione @edoakes There's actually one issue now. Since our export / harvest time is 10 seconds, the shutdown takes very long time (it could be maximum 10 seconds in the worst case). Any good idea how to improve this?

**Comment by ashione** [comment]
> @ashione @edoakes There's actually one issue now. Since our export / harvest time is 10 seconds, the shutdown takes very long time (it could be maximum 10 seconds in the worst case). Any good idea how to improve this?

boost timer may be helpful for stats RunLoop thread, but it's too many hack code to minimize changes in opencensus patch.

**Comment by rkooo567** [comment]
@ashione I actually made each loop run every 0.5 second and invoke export or flush only when necessary. Please take a look at code. I addressed all other comments of you too.

File: src/ray/stats/stats.h (hunk 2 above)
Comment (ashione) [reply src/ray/stats/stats.h:111]: > so that we can flush and export right? I think that's a good idea.

It can flush metric at last emit moment.

**Comment by ashione** [comment]
> @ashione I actually made each loop in a thread run every 0.5 second and invoke export or flush only when necessary. Please take a look at code. I addressed all other comments of you too. New code will lose the last batch of metrics. (I can make it work, but it will be pretty ugly).

Sure, it's acceptable to lose last batch metrics.

**Review by ashione — APPROVED** [review]

**Pull request merged**